### PR TITLE
Make make_dual redispatch

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68631
* #68629
* __->__ #68630
* #68628

Constraints:
1) (functorch) if all the inputs to an op have requires_grad=False and don't have tangents, then their VariableType
    kernel should be a no-op i.e., behave like a redispatch. This is due to functorch's DynamicLayerStack
   having the autograd key by default (which is so that transformations like vmap) still work with autograd
2) (inference mode) inference tensors in inference mode will call straight into the kernel, we should still do something sensible   
    inside even if we normally wouldn't redispatch into it.
3) ~Should support potential application of interposition below autograd: `nn.Parameter` is a example of subclassing where the subclass  
    is not preserved when an operation is performed. There is an exception though: we want calling `make_dual` on a 
    `nn.Parameter` to preserve its parameterness.~
4) Should avoid calls to shallow_copy_and_detach to avoid spurious calls into `__python_dispatch__`. 

This PR:
- does not redispatch to `make_dual` from its `ADInplaceOrView` kernel to satisfy (1)
- calls into `alias` from the kernel in the native namespace so that behavior is consistent with other views in inference mode to satisfy (2)
- discussion of (3). We still wouldn't be able to directly override `make_dual` below autograd. In this PR, instead of not redispatching at all, we choose to redispatch into `at::alias` so that one can override `make_dual`. The side effect is that one would not be able to distinguish calls between the two, which can be problematic (though a straightforward but hacky solution would be to create a new `at::alias_for_make_dual` that would allow users to distinguish) the two. This isn't ideal but seems to be the simplest way to satisfy (3). We don't pursue that hacky solution here.
- (4) is satisfied because we remove calls to `shallow_copy_and_detach`

<details>
<summary> A potentially less hacky but more involved solution? (WIP) </summary>

Realizing that make_dual is more like requires_grad, perhaps it shouldn't be autograd explicit? Make make_dual a composite or python-only construct. i.e., it would be a view on the primal followed by something to the effect of primal.set_fw_grad(tangent).
         
Additional constraints:
5) make_dual needs to be backward-differentiable (I can't think of any applications yet becuase
   technically as a high-order function, jvp's input is the tangent only, "detach" is not applied on
   the tangent, so one would still be able to propagate gradients through it).
6) set_fw_grad needs to raise an error if there is a layout mismatch and base is a forward-differnentiable view

Possible plan
- (6) implies that a plain view would not suffice. We need a `detach`-like operation to ensure that set_fw_grad 
  knows the view is not forward differentiable.
- (5) implies that is this (new) `detach` would need to be backward differentiable (API TBD).
- (3) is no longer relevant because make_dual is no longer autograd explicit, but perhaps this new detach should behave like the current one? There is a lot of logic to replicate for detach, so this may be hard.
- (1) is satisfied if we use current detach logic, i.e., , and (4) is trivial.

I'm not convinced that this is the right solution either, because in the end does (3) still work?

 </details>

Differential Revision: [D32899679](https://our.internmc.facebook.com/intern/diff/D32899679)

### CODE REVIEW

**Comment by soulitzer** [comment]
From offline discussion: add check that verifies that going into native function can only be done in inference mode.

File: aten/src/ATen/native/AutogradComposite.cpp [hunk 1]
```diff
@@ -4,12 +4,8 @@
 namespace at {
 namespace native {
 
-// _make_dual is actually implemented in VariableTypeManual.cpp, this is just a stub
```
Comment (albanD) [inline aten/src/ATen/native/AutogradComposite.cpp:7]: I think we could leave a comment here saying that this should only be called in inference mode.

**Review by albanD — APPROVED** [review]
Are you sure about 3. ?
Or was it dropped and the description wasn't updated?

This is a good fix for the inference. We still need to make a choice if we want make_dual/unpack_dual/detach to be composite implicit or not.

**Comment by soulitzer** [comment]
> Are you sure about 3. ?

Ehhh, probably a good idea to cross it out from this list, since we don't really address it here, but good to support eventually I think.

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899679).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899679).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899679).
